graphula-core/src/Graphula.hs

```diff
@@ -193,14 +201,36 @@ instance (MonadIO m, Applicative n, MonadIO n) => MonadGraphulaFrontend (Graphul
         Just key -> do
           existingKey <- get key
           whenNothing existingKey $ do
-            existingUnique <- checkUnique n
+            let updated = embedKey key n
+            existingUnique <- checkUnique updated
             whenNothing existingUnique $ do
-              insertKey key n
+              insertKey key updated
               getEntity key
   remove key = do
     RunDB runDB <- ask
     lift . runDB $ delete key
 
+-- | Embed generated key components in record
+--
+-- @'insertKey'@ doesn't do this for us, and we want to update the
+-- record before doing a @'checkUnique'@ anyway.
+--
+embedKey :: PersistEntity a => Key a -> a -> a
+embedKey key n = if isNothing $ entityPrimary def
+  then n
+  else fromRight n $ fromPersistValues $ zipWith
+    replaceField
+    (fieldHaskell <$> entityFields def)
+    (toPersistValue <$> toPersistFields n)
+ where
+  def = entityDef $ Just n
+
+  keyMap :: [(HaskellName, PersistValue)]
+  keyMap = zip (fieldHaskell <$> entityKeyFields def) (keyToValues key)
+
+  replaceField :: HaskellName -> PersistValue -> PersistValue
+  replaceField name value = fromMaybe value $ lookup name keyMap
```

Comment:
Most of this is modified from [`entityValues` in `persistent`](https://www.stackage.org/haddock/lts-14.12/persistent-2.9.2/Database-Persist-Class.html#v:entityValues).